index/internal/indexer/indexer.go

```diff
@@ -0,0 +1,302 @@
+		Deps               collections.Set[label.Label] // dependencies on other targets
+	}
+)
+
+type IndexingResult struct {
+	// Headers mapping to exactly one Bazel rule
+	HeaderToRule map[string]label.Label
+	// Headers defined in multiple rules
+	Ambiguous map[string]collections.Set[label.Label]
+}
+
+// Process list of modules to create an unfiorm index mapping header to exactly one rule that provides their definition.
+// In case if multiple modules define same headers might try to select one that behaves as clousers over remaining ambigious rules.
+func CreateHeaderIndex(modules []Module) IndexingResult {
+	// headersMapping will store header paths to a collections.Set of Labels.
+	headersMapping := map[string]*collections.Set[label.Label]{}
+	for _, module := range modules {
+		for _, target := range module.Targets {
+			// Create a targetLabel for the target using the module repository.
+			// It's required to correctly map external module to sources found possibly in other rules
+			targetLabel := label.New(module.Repository, target.Name.Pkg, target.Name.Name)
+			if shouldExcludeTarget(targetLabel) {
+				continue
+			}
+
+			// Normalize headers and add to mapping
+			for hdr := range target.Hdrs {
+				normalizedPath := normalizeHeaderPath(hdr.Name, *target)
+				if shouldExcludeHeader(normalizedPath) {
+					continue
+				}
+				if _, exists := headersMapping[normalizedPath]; !exists {
+					headersMapping[normalizedPath] = &collections.Set[label.Label]{}
+				}
+				headersMapping[normalizedPath].Add(targetLabel)
+			}
+		}
+	}
+
+	// Partition the headers into non-conflicting (exactly one label) and ambiguous (multiple labels).
+	headerToRule := make(map[string]label.Label)
+	ambiguous := make(map[string]collections.Set[label.Label])
+	for path, labels := range headersMapping {
+		switch len(*labels) {
+		case 1:
+			// Extract the only label in the collections.Set.
+			for l := range *labels {
+				headerToRule[path] = l
+				break
+			}
+		default:
+			ambiguous[path] = *labels
+		}
+	}
+
+	return IndexingResult{
+		HeaderToRule: headerToRule,
+		Ambiguous:    ambiguous,
+	}
+}
+
+// Writes the mapping of IndexingResult.HeaderToRule to disk in JSON format.
+// Labels are stored as renered strings
+func (result IndexingResult) WriteToFile(outputFile string) {
+	mappings := make(map[string]string, len(result.HeaderToRule))
+	for hdr, lbl := range result.HeaderToRule {
+		mappings[hdr] = lbl.String()
+	}
+
+	data, err := json.MarshalIndent(mappings, "", "  ")
+	if err != nil {
+		log.Fatalf("Failed to serialize header index to JSON: %v", err.Error())
+	}
+
+	os.MkdirAll(filepath.Dir(outputFile), 0644)
+	if err := os.WriteFile(outputFile, data, 0644); err != nil {
+		log.Fatalf("Failed to write index file: %v", err.Error())
+	}
+}
+
+// Prints to stdout detailed information about headers with resolved mappings and the ambigious header definitions.
+func (result IndexingResult) Show() {
```

Comment:
Maybe implement `String() string` instead? Then it's a lot more testable, and the caller can decide whether they want to use the `log` package or something else.